modules/python/nitf/source/nitro.i

```diff
@@ -549,64 +549,70 @@
         /* TODO somehow get the NUMBITSPERPIXEL in the future */
         nitf_Uint8 **buf = NULL;
         PyObject* result = NULL;
-        int i, padded, rowSkip, colSkip, subimageSize;
-        nitf_SubWindow* newWindow = nitf_SubWindow_construct(error);
-        
+        int i, padded, rowSkip, colSkip, subimageSize, bandsMalloced;
+
+        bandsMalloced = 0;
+
         rowSkip = window->downsampler ? window->downsampler->rowSkip : 1;
         colSkip = window->downsampler ? window->downsampler->colSkip : 1;
-        subimageSize = (window->numRows/rowSkip) * (window->numCols/colSkip) * NITF_NBPP_TO_BYTES(nbpp);
-        
-        buf = (nitf_Uint8**) NITF_MALLOC(sizeof(nitf_Uint8*));
+        subimageSize = (window->numRows/rowSkip) * (window->numCols/colSkip) * nitf_ImageIO_pixelSize(reader->imageDeblocker);
+
+        buf = (nitf_Uint8**) NITF_MALLOC(sizeof(nitf_Uint8*) * window->numBands);
         if (!buf)
         {
             PyErr_NoMemory();
             goto CATCH_ERROR;
         }
-        
-        /* copy the window */
-        newWindow->downsampler = window->downsampler;
-        newWindow->startRow = window->startRow;
-        newWindow->startCol = window->startCol;
-        newWindow->numRows = window->numRows;
-        newWindow->numCols = window->numCols;
-        newWindow->numBands = 1;
-        
-        result = PyList_New(window->numBands);
-        for (i = 0; i < window->numBands; i++)
+
+        for (i = 0; i < window->numBands; ++i)
         {
-            PyObject* buffObj = PyBuffer_New(subimageSize * sizeof(nitf_Uint8));
-            if (!buffObj) goto CATCH_ERROR;
-            buffObj->ob_type->tp_as_buffer->bf_getwritebuffer(buffObj, 0, (void **)&buf[0]);
-            
-            newWindow->bandList = &window->bandList[i];
-            if (!nitf_ImageReader_read(reader, newWindow, buf, &padded, error))
+            buf[i] = (nitf_Uint8*) NITF_MALLOC(sizeof(nitf_Uint8) * subimageSize);
+            if (!buf[i])
             {
-                nitf_Error_print(error, stderr, "Read failed");
+                PyErr_NoMemory();
                 goto CATCH_ERROR;
             }
+
+            ++bandsMalloced;
+        }
+
+        result = PyList_New(window->numBands);
+
+        if (!nitf_ImageReader_read(reader, window, buf, &padded, error))
+        {
+            nitf_Error_print(error, stderr, "Read failed");
+            goto CATCH_ERROR;
+        }
+
+        for (i = 0; i < window->numBands; i++)
+        {
+            PyObject* buffObj = PyBuffer_FromMemory(buf[i], subimageSize * sizeof(nitf_Uint8));
             PyList_SetItem(result, i, buffObj);
         }
-        /* destroy temp window */
-        newWindow->bandList = NULL;
-        newWindow->downsampler = NULL;
-        nitf_SubWindow_destruct(&newWindow);
-        
-        NITF_FREE(buf);
+
+        //The PyList now owns the memory, so not freeing the buffer here
```

Comment:
So the PyList owns `buf[0]`, `buf[1]`, through `buf[numBands-1]` but nothing actually owns the `buf` pointer itself does it?  That is, I think you still need the `NITF_FREE(buf)` don't you?